### PR TITLE
Endpoint factory manager destroy race condition

### LINKED ISSUE #2585: C++ race condition with Communicator::stringToProxy and Communicator::destroy
While porting Swift to use async/await I got this crash

```
* thread #2, queue = 'com.apple.root.default-qos.cooperative', stop reason = EXC_BAD_ACCESS (code=1, address=0x5f)
  * frame #0: 0x0000000105081c6c TestDriver`std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::__is_long[abi:ue170006](this="") const at string:1734:33
    frame #1: 0x0000000105081c1c TestDriver`std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::__get_pointer[abi:ue170006](this="") const at string:1869:17
    frame #2: 0x0000000105081bdc TestDriver`std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>::data[abi:ue170006](this="") const at string:1559:73
    frame #3: 0x00000001050ca1c4 TestDriver`std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::operator<<[abi:ue170006]<char, std::__1::char_traits<char>, std::__1::allocator<char>>(__os=0x000000016af15020, __str="") at ostream:1093:56
    frame #4: 0x00000001051272c4 TestDriver`IceInternal::EndpointI::initWithOptions(this=0x000060000166c418, args=size=2) at EndpointI.cpp:49:17
    frame #5: 0x0000000105138c54 TestDriver`IceInternal::IPEndpointI::initWithOptions(this=0x000060000166c418, args=size=2, oaEndpoint=false) at IPEndpointI.cpp:378:16
    frame #6: 0x000000010546ed10 TestDriver`IceInternal::TcpEndpointFactory::create(this=0x00006000024743d8, args=size=2, oaEndpoint=false) const at TcpEndpointI.cpp:364:12
    frame #7: 0x000000010512328c TestDriver`IceInternal::EndpointFactoryManager::create(this=0x0000600000474398, str="tcp -p 12010", oaEndpoint=false) const at EndpointFactoryManager.cpp:108:35
    frame #8: 0x00000001053fdeb8 TestDriver`IceInternal::ReferenceFactory::create(this=0x0000600001f74018, str="test:tcp -p 12010", propertyPrefix="") at ReferenceFactory.cpp:469:74
    frame #9: 0x00000001050989e8 TestDriver`Ice::Communicator::_stringToProxy(this=0x0000600003f4c658, s="test:tcp -p 12010") const at Communicator.cpp:84:43
    frame #10: 0x00000001056e37d0 TestDriver`std::__1::optional<Ice::ObjectPrx> Ice::Communicator::stringToProxy<Ice::ObjectPrx, true>(this=0x0000600003f4c658, str="test:tcp -p 12010") const at Communicator.h:95:30
    frame #11: 0x00000001056e35c8 TestDriver`-[ICECommunicator stringToProxy:error:](self=0x0000600002a703a0, _cmd="stringToProxy:error:", str="test:tcp -p 12010", error=0x000000016af16a70) at Communicator.mm:51:39
    frame #12: 0x0000000104fab6bc TestDriver`closure #1 in CommunicatorI.stringToProxyImpl<ProxyImpl>(self=0x0000600000e50d80, str="test:tcp -p 12010") at CommunicatorI.swift:251:46
    frame #13: 0x0000000104fad9c4 TestDriver`partial apply for closure #1 in CommunicatorI.stringToProxyImpl<A>(_:) at <compiler-generated>:0
    frame #14: 0x00000001a2d1c54c libswiftObjectiveC.dylib`ObjectiveC.autoreleasepool<τ_0_0>(invoking: () throws -> τ_0_0) throws -> τ_0_0 + 64
    frame #15: 0x0000000104fab14c TestDriver`CommunicatorI.stringToProxyImpl<ProxyImpl>(str="test:tcp -p 12010", self=0x0000600000e50d80) at CommunicatorI.swift:250:20
    frame #16: 0x0000000104fa6ca8 TestDriver`CommunicatorI.stringToProxy(str="test:tcp -p 12010", self=0x0000600000e50d80) at CommunicatorI.swift:48:13
    frame #17: 0x0000000104fabb64 TestDriver`protocol witness for Communicator.stringToProxy(_:) in conformance CommunicatorI at <compiler-generated>:0
    frame #18: 0x00000001057c0358 TestDriver`closure #1 in allTests(output=0x0000600002860230, comm=0x0000600000e50d80, ref="test:tcp -p 12010") at AllTests.swift:60:36
    frame #19: 0x00000001057c0d44 TestDriver`partial apply for closure #1 in allTests(_:) at <compiler-generated>:0
    frame #20: 0x0000000104fa2214 TestDriver`thunk for @escaping @callee_guaranteed @Sendable @async () -> (@out A) at <compiler-generated>:0
    frame #21: 0x0000000104fa2370 TestDriver`partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> (@out A) at <compiler-generated>:0
```

The code in question was running stringToProxy and destroy at the same time.

```
Task {
    try communicator.stringToProxy(ref)
}
communicator.destory()
```

### PR DESCRIPTION
**Pull request by pepone**
Fix #2585

### CODE REVIEW

**Comment by pepone** [comment]
> Would be nice to fix the Swift test that found this to put the proxy creating back inside the task.

This can result in CommunicatorDestroyedException when the proxy is created.

**Comment by externl** [comment]
> > Would be nice to fix the Swift test that found this to put the proxy creating back inside the task.
> 
> This can result in CommunicatorDestroyedException when the proxy is created.



> > Would be nice to fix the Swift test that found this to put the proxy creating back inside the task.
> 
> This can result in CommunicatorDestroyedException when the proxy is created.

That should be fine, just as long as it doesn't crash. We can catch the `CommunicatorDestroyedException` and ignore in the task.

**Comment by pepone** [comment]
> That should be fine, just as long as it doesn't crash. We can catch the CommunicatorDestroyedException and ignore in the task. That comment is only there because of the race condition.

Isn't the point of this test that the pingAsync creates a connection, if the proxy is not created what are we testing here?

**Comment by externl** [comment]
> > That should be fine, just as long as it doesn't crash. We can catch the CommunicatorDestroyedException and ignore in the task. That comment is only there because of the race condition.
> 
> Isn't the point of this test that the pingAsync creates a connection, if the proxy is not created what are we testing here?

Ok I see your point. Let's leave the test as it is!

File: cpp/src/Ice/EndpointFactoryManager.cpp [hunk 1]
```diff
@@ -31,7 +31,11 @@ IceInternal::EndpointFactoryManager::initialize() const
 void
 IceInternal::EndpointFactoryManager::add(const EndpointFactoryPtr& factory)
 {
-    lock_guard lock(_mutex); // TODO: Necessary?
+    lock_guard lock(_mutex);
+    if (_destroyed)
+    {
+        throw CommunicatorDestroyedException(__FILE__, __LINE__);
```
Comment (bernardnormier) [inline cpp/src/Ice/EndpointFactoryManager.cpp:37]: Prefer `{}` for constructors.

File: cpp/src/Ice/EndpointFactoryManager.cpp [hunk 2]
```diff
@@ -89,8 +97,12 @@ IceInternal::EndpointFactoryManager::create(const string& str, bool oaEndpoint)
 
     EndpointFactoryPtr factory;
     {
-        lock_guard lock(_mutex); // TODO: Necessary?
-
+        // Hold the mutex while looking up the factory, destroy can be called concurrently.
```
Comment (bernardnormier) [inline cpp/src/Ice/EndpointFactoryManager.cpp:100]: I don't see how this PR solves the race condition.

When you call the factory later on in this function, it's outside the lock, and the factory code itself is not thread-safe. In particular, we could be destroying the factory (setting its _instance to nullptr - see TcpEndpointFactory) and using this factory concurrently.

In my view, the only proper way to solve this issue is to use weak_ptr.

File: csharp/src/Ice/Internal/EndpointFactoryManager.cs [hunk 3]
```diff
@@ -24,6 +24,11 @@ public void add(EndpointFactory factory)
     {
         lock (_mutex)
         {
+            if (_destroyed)
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/EndpointFactoryManager.cs:27]: For C# and Java, it would be much simpler to remove destroy/_destroyed. destroy is about breaking cycles in C++, and we don't need that in C# and Java. On the contrary, it makes the code much more complicated and (like here) incorrect since it makes race conditions possible in the endpoint factory implementations.

File: cpp/src/Ice/EndpointFactoryManager.cpp (hunk 2 above)
Comment (pepone) [reply cpp/src/Ice/EndpointFactoryManager.cpp:90]: Updated the code to use a weak_ptr.

File: cpp/src/Ice/ProtocolInstance.cpp [hunk 4]
```diff
@@ -25,9 +25,9 @@ IceInternal::ProtocolInstance::ProtocolInstance(
     const string& protocol,
     bool secure)
     : _instance(getInstance(com)),
-      _traceLevel(_instance->traceLevels()->network),
-      _traceCategory(_instance->traceLevels()->networkCat),
-      _properties(_instance->initializationData().properties),
+      _traceLevel(_instance.lock()->traceLevels()->network),
```
Comment (pepone) [inline cpp/src/Ice/ProtocolInstance.cpp:28]: This shouldn't be a problem the `com` parameter has a strong reference to the instance.

**Review by bernardnormier — COMMENTED** [review]
The C# and Java code looks fine (but see comments).

I don't think the C++ code is correct - it creates a leak. That's because EndpointFactory holds a strong/never cleared ProtocolInstancePtr.

The base class ProtocolInstance is fine (no strong reference to Communicator/Instance). But we have derived ProtocolInstances with all kind of other data members. For example:

https://github.com/zeroc-ice/ice/blob/2846e40a8ea07d870d762d89b2075071ab04aba2/cpp/src/Ice/SSL/SSLInstance.h#L22
https://github.com/zeroc-ice/ice/blob/2846e40a8ea07d870d762d89b2075071ab04aba2/cpp/src/Ice/SSL/SSLEngine.h#L59

In other words, the EndpointFactory need a weak_ptr for its ProtocolInstance.

File: cpp/src/Ice/EndpointFactoryManager.cpp [hunk 5]
```diff
@@ -49,8 +48,7 @@ IceInternal::EndpointFactoryManager::add(const EndpointFactoryPtr& factory)
 EndpointFactoryPtr
 IceInternal::EndpointFactoryManager::get(int16_t type) const
 {
-    lock_guard lock(_mutex); // TODO: Necessary?
-
+    lock_guard lock(_mutex);
```
Comment (pepone) [inline cpp/src/Ice/EndpointFactoryManager.cpp:51]: Previously the lock was required because "destroy" clears the map, and "get" can be iterating it concurrently.

Now with the removal of destroy, I think it is still required in case plug-ins load endpoint-factories, while a remote call access the factories from a separate thread. Not very likely but seems safer to keep the lock.

Comment (bernardnormier) [reply cpp/src/Ice/EndpointFactoryManager.cpp:51]: We should clarify if plugin initialization can occur concurrently while other threads are using/calling the communicator.

In my view, the answer should be No. I don't see a use-case for allowing such concurrent initialization.

File: cpp/src/Ice/SSL/SSLEndpointI.h [hunk 6]
```diff
@@ -69,8 +69,6 @@ namespace Ice::SSL
     public:
         EndpointFactoryI(const InstancePtr&, std::int16_t);
 
-        void destroy() final;
```
Comment (bernardnormier) [inline cpp/src/Ice/SSL/SSLEndpointI.h:72]: Make _sslInstance const below.

File: cpp/test/Ice/background/EndpointFactory.h [hunk 7]
```diff
@@ -16,7 +16,6 @@ class EndpointFactory final : public IceInternal::EndpointFactory, public std::e
     std::string protocol() const final;
     IceInternal::EndpointIPtr create(std::vector<std::string>&, bool) const final;
     IceInternal::EndpointIPtr read(Ice::InputStream*) const final;
-    void destroy() final;
```
Comment (bernardnormier) [inline cpp/test/Ice/background/EndpointFactory.h:19]: _factory should be const?

**Review by bernardnormier — APPROVED** [review]
Looks good.

**Review by externl — APPROVED** [review]
